addons/sale_margin/models/sale_order.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, fields, models
+import openerp.addons.decimal_precision as dp
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    margin = fields.Float(compute='_product_margin', digits=dp.get_precision('Product Price'), store=True)
+    purchase_price = fields.Float(string='Cost Price', digits=dp.get_precision('Product Price'))
+
+    @api.multi
+    @api.onchange('product_id', 'product_uom_qty')
+    def product_id_change_margin(self):
+        for line in self:
```

Comment:
@mart-e ok, But i don't understand the purpose of using @api.multi and @api.onchange both decorators ?. Moreover, onchange method contain only one record, so no need to iterate over 'self' .